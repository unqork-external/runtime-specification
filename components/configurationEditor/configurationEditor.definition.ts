import { Default, Description, DiscriminatorValue, Property, Required } from '@tsed/schema'

import { BaseSettingDefinition } from '../../../designer-specification'
import { BaseComponentDefinition } from '../../base-component-interface'
import { StandardArrayNestable } from '../../nestables'

@DiscriminatorValue('configurationEditor')
export class ConfigurationEditorComponentDefinition extends BaseComponentDefinition {
  @Required()
  type: 'configurationEditor' = 'configurationEditor' as const

  @Required()
  @Default({})
  @Description('The draft state of the config that is currently being used by the config editor.')
  currentConfig: Record<string, any> = {}

  @Required()
  @Default({})
  @Description('The slice of state designated holding an asset type paired with its BaseSettingDefinition.')
  assetsSettingsSpecs: Record<string, BaseSettingDefinition> = {}

  @Required()
  @Default(false)
  @Description('Boolean to determine if the config areas should be pulled from the assets.')
  useSpecFromAssets: boolean = false

  @Description('Child ids for nestable references')
  childIds: string[] = []

  @Required()
  @Property()
  @Default(false)
  @Description('Boolean to determine if default styles should be used.')
  useLegacy: boolean = false

  @Description('Nestable information for configuration editor')
  nestables = { childIds: new StandardArrayNestable() }
}
