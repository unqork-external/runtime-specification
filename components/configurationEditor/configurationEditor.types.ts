import { Description, Required } from '@tsed/schema'

export class ExternalComponentRef {
  @Required()
  @Description('Key of external component')
  key: string

  @Required()
  @Description('Type of external component')
  type: string
}
