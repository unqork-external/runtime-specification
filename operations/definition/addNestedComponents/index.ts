export * from './addNestedComponents.model'
export * from './addNestedComponents.options'
