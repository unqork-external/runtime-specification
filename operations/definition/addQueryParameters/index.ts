export * from './addQueryParameters.model'
export * from './addQueryParameters.options'
