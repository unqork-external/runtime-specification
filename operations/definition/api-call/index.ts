export * from './apiCall.type'
export * from './method.type'
export * from './serviceType.type'
export * from './apiCall.model'
export * from './apiCall.options'
