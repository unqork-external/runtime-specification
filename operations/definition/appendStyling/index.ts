export * from './appendStyling.model'
export * from './appendStyling.options'
