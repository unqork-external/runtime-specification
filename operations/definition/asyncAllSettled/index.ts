export * from './asyncAllSettled.model'
export * from './asyncAllSettled.options'
