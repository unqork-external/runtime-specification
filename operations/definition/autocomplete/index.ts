export * from './autocomplete.model'
export * from './autocomplete.options'
