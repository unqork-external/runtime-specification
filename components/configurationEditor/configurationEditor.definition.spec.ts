import { ConfigurationEditorComponentDefinition } from './configurationEditor.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('ConfigurationEditor Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(ConfigurationEditorComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
    expect(schema.definitions).toMatchSnapshot()
  })
})
