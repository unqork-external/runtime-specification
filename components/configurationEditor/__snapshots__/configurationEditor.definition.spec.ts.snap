// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationEditor Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
  "currentConfig",
  "assetsSettingsSpecs",
  "useSpecFromAssets",
]
`;

exports[`ConfigurationEditor Contract Tests should ensure the contract is clear 2`] = `
{
  "assetsSettingsSpecs": {
    "default": {},
    "description": "The slice of state designated holding an asset type paired with its BaseSettingDefinition.",
    "type": "object",
  },
  "childIds": {
    "description": "Child ids for nestable references",
    "type": "array",
  },
  "currentConfig": {
    "default": {},
    "description": "The draft state of the config that is currently being used by the config editor.",
    "type": "object",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "nestables": {
    "description": "Nestable information for configuration editor",
    "type": "object",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/StylingModel",
  },
  "type": {
    "const": "configurationEditor",
    "description": "Name of the type of component.",
    "examples": [
      "configurationEditor",
    ],
    "minLength": 1,
    "type": "string",
  },
  "useSpecFromAssets": {
    "default": false,
    "description": "Boolean to determine if the config areas should be pulled from the assets.",
    "type": "boolean",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Current value of the component",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;

exports[`ConfigurationEditor Contract Tests should ensure the contract is clear 3`] = `
{
  "InputRef": {
    "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
    "properties": {
      "alias": {
        "type": "string",
      },
      "propertyPath": {
        "type": "string",
      },
      "required": {
        "type": "boolean",
      },
      "silent": {
        "type": "boolean",
      },
      "targetKey": {
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "required": [
      "targetKey",
    ],
    "type": "object",
  },
  "PackageModel": {
    "description": "The PackageModel describes how the Vega Runtime will resolve dependencies",
    "properties": {
      "type": {
        "description": "Type of package",
        "type": "string",
      },
      "url": {
        "description": "URL of package",
        "type": "string",
      },
      "version": {
        "description": "Version of package",
        "type": "string",
      },
    },
    "type": "object",
  },
  "StylingModel": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "Validation": {
    "description": "Holds the validation rules and the most current validation results for the component",
    "properties": {
      "isDirty": {
        "description": "Whether or not field has had user interaction. It will be set to true on user interaction.",
        "type": "boolean",
      },
      "result": {
        "$ref": "#/definitions/ValidationResult",
      },
      "rules": {
        "description": "The set of rules to restrict or define the accepted data for the current component. These are the rules to run for validation of the data input.",
        "examples": [
          "You can use it to define the max length of characters in an input field with a custom error message. rules: { required: { type: ValidationRuleTypes.MAX_LENGTH, validation: { maxLength: 5, }, errorMessage: 'Max length is 5', }, },",
        ],
        "type": "object",
      },
    },
    "type": "object",
  },
  "ValidationError": {
    "properties": {
      "message": {
        "description": "The error message",
        "type": "string",
      },
      "type": {
        "description": "Type of the error",
        "enum": [
          "required",
          "maxLength",
          "minLength",
          "arrayMinLength",
          "arrayMaxLength",
          "pattern",
          "min",
          "max",
          "mask",
          "number",
          "dateinput",
          "isBeforeDate",
          "isAfterDate",
          "customError",
          "isEmail",
        ],
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "type",
    ],
    "type": "object",
  },
  "ValidationResult": {
    "properties": {
      "errors": {
        "items": {
          "$ref": "#/definitions/ValidationError",
        },
        "type": "array",
      },
      "isValid": {
        "type": [
          "null",
          "boolean",
        ],
      },
    },
    "type": "object",
  },
}
`;
